persist/fs/commitlog/reader.go

```diff
@@ -56,34 +61,84 @@ type commitLogReader interface {
 	Close() error
 }
 
+type readResponse struct {
+	series     Series
+	datapoint  ts.Datapoint
+	unit       xtime.Unit
+	annotation ts.Annotation
+	resultErr  error
+}
+
+type decoderArg struct {
+	bytes checked.Bytes
+	err   error
+}
+
+type readerMetadata struct {
+	sync.RWMutex
+	decoder encoding.Decoder
+	lookup  map[uint64]Series
+	wgs     map[uint64]*sync.WaitGroup
+}
+
 type reader struct {
-	opts            Options
-	bytesPool       pool.CheckedBytesPool
-	chunkReader     *chunkReader
-	sizeBuffer      []byte
-	dataBuffer      []byte
-	logDecoder      encoding.Decoder
-	metadataDecoder encoding.Decoder
-	metadataLookup  map[uint64]Series
+	opts          Options
+	numConc       int
+	bytesPool     pool.CheckedBytesPool
+	chunkReader   *chunkReader
+	dataBuffer    []byte
+	logDecoder    encoding.Decoder
+	decoderBufs   []chan decoderArg
+	outBufs       []chan readResponse
+	cancelCtx     context.Context
+	cancelFunc    context.CancelFunc
+	shutdownChan  chan error
+	metadata      readerMetadata
+	nextIndex     int
+	hasBeenOpened bool
 }
 
 func newCommitLogReader(opts Options) commitLogReader {
 	decodingOpts := opts.FilesystemOptions().DecodingOptions()
-	return &reader{
-		opts:            opts,
-		bytesPool:       opts.BytesPool(),
-		chunkReader:     newChunkReader(opts.FlushSize()),
-		sizeBuffer:      make([]byte, binary.MaxVarintLen64),
-		logDecoder:      msgpack.NewDecoder(decodingOpts),
-		metadataDecoder: msgpack.NewDecoder(decodingOpts),
-		metadataLookup:  make(map[uint64]Series),
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+
+	numConc := opts.ReadConcurrency()
+	decoderBufs := []chan decoderArg{}
+	for i := 0; i < numConc; i++ {
+		decoderBufs = append(decoderBufs, make(chan decoderArg, decoderInBufChanSize))
 	}
+	outBufs := []chan readResponse{}
+	for i := 0; i < numConc; i++ {
+		outBufs = append(outBufs, make(chan readResponse, decoderInBufChanSize))
+	}
+
+	reader := &reader{
+		opts:         opts,
+		numConc:      numConc,
+		bytesPool:    opts.BytesPool(),
+		chunkReader:  newChunkReader(opts.FlushSize()),
+		logDecoder:   msgpack.NewDecoder(decodingOpts),
+		decoderBufs:  decoderBufs,
+		outBufs:      outBufs,
+		cancelCtx:    cancelCtx,
+		cancelFunc:   cancelFunc,
+		shutdownChan: make(chan error),
+		metadata: readerMetadata{
+			decoder: msgpack.NewDecoder(decodingOpts),
```

Comment:
you create a new metadata decoder at the start of `readLoop()`, do you still need this?

Reply:
I think you're right, I'll get rid of it